src/ray/gcs/job_state_accessor.h

```diff
@@ -0,0 +1,71 @@
+#ifndef RAY_GCS_JOB_STATE_ACCESSOR_H
+#define RAY_GCS_JOB_STATE_ACCESSOR_H
+
+#include "ray/common/id.h"
+#include "ray/gcs/callback.h"
+#include "ray/gcs/subscription_executor.h"
+#include "ray/gcs/tables.h"
+
+namespace ray {
+
+namespace gcs {
+
+class RedisGcsClient;
+
+/// \class JobStateAccessor
+/// JobStateAccessor class encapsulates the implementation details of
+/// reading or writing or subscribing of job's information (immutable fields which
+/// determined at submission time, and mutable fields which determined at runtime).
+class JobStateAccessor {
+ public:
+  explicit JobStateAccessor(RedisGcsClient &client_impl);
+
+  ~JobStateAccessor() {}
+
+  /// Register a job to GCS asynchronously.
+  ///
+  /// \param data_ptr The job that will be registered to GCS.
+  /// \param callback Callback that will be called after job has been registered
+  /// to GCS.
+  /// \return Status
+  Status AsyncRegister(const std::shared_ptr<JobTableData> &data_ptr,
```

Comment:
what about naming this `AsyncAddJob`, whose meaning is clearer than `register`?
Also, we can just pass in `job_id`, `timestamp`, `node_manager_address` and `driver_pid`, because `is_dead` is already `false` and we want to hide the implementation details about `table`.

Reply:
> what about naming this `AsyncAddJob`, whose meaning is clearer than `register`?

I think `add` it's ok.  But `add` it's more suitable for adding data, what if gcs is a service not only a storage in future? There might be some other logic not only adding data.

> Also, we can just pass in `job_id`, `timestamp`, `node_manager_address` and `driver_pid`, because `is_dead` is already `false` and we want to hide the implementation details about `table`.

I think it's better use a struct rather than list all parameters.  Because if we add a parameter in future, we won't need change the interface.